tests/lib/ocpp/v201/test_smart_charging_handler.cpp

```diff
@@ -1678,4 +1683,394 @@ TEST_F(SmartChargingHandlerTestFixtureV201, K05FR02_RequestStartTransactionReque
+    };
+    double deltaChanged = 0.2;
+    auto source = ChargingLimitSourceEnum::Other;
+
+    auto resp = handler.handle_external_limits_changed(new_limit, deltaChanged, source, DEFAULT_EVSE_ID);
+
+    ASSERT_THAT(resp.has_value(), testing::IsTrue());
+    ASSERT_THAT(resp->chargingSchedule.has_value(), testing::IsFalse());
+}
+
+TEST_F(SmartChargingHandlerTestFixtureV201, K12FR04_HandleExternalLimitsChanged_NotificationIncludesSource) {
+    const auto& limit_change_cv = ControllerComponentVariables::LimitChangeSignificance;
+    device_model->set_value(limit_change_cv.component, limit_change_cv.variable.value(), AttributeEnum::Actual, "0.1",
+                            "test");
+
+    ConstantChargingLimit new_limit = {
+        .limit = 100.0,
+        .charging_rate_unit = ChargingRateUnitEnum::A,
+    };
+    double deltaChanged = 0.2;
+    auto source = ChargingLimitSourceEnum::SO;
+
+    auto resp = handler.handle_external_limits_changed(new_limit, deltaChanged, source, DEFAULT_EVSE_ID);
+
+    ASSERT_THAT(resp.has_value(), testing::IsTrue());
+    ASSERT_THAT(resp->chargingLimit.chargingLimitSource, testing::Eq(source));
+}
+
+TEST_F(SmartChargingHandlerTestFixtureV201, K12FR04_HandleExternalLimitsChanged_ThrowsIfSourceIsCSO) {
+    const auto& limit_change_cv = ControllerComponentVariables::LimitChangeSignificance;
+    device_model->set_value(limit_change_cv.component, limit_change_cv.variable.value(), AttributeEnum::Actual, "0.1",
+                            "test");
+
+    ConstantChargingLimit new_limit = {
+        .limit = 100.0,
+        .charging_rate_unit = ChargingRateUnitEnum::A,
+    };
+    double deltaChanged = 0.2;
+    auto source = ChargingLimitSourceEnum::CSO;
+
+    EXPECT_THROW(handler.handle_external_limits_changed(new_limit, deltaChanged, source, DEFAULT_EVSE_ID),
+                 std::invalid_argument);
+}
+
+TEST_F(SmartChargingHandlerTestFixtureV201, K12_HandleExternalLimitsChanged_RequestHasGivenEvseId) {
+    const auto& limit_change_cv = ControllerComponentVariables::LimitChangeSignificance;
+    device_model->set_value(limit_change_cv.component, limit_change_cv.variable.value(), AttributeEnum::Actual, "0.1",
+                            "test");
+
+    auto evse_id = 12;
+
+    ConstantChargingLimit new_limit = {
+        .limit = 100.0,
+        .charging_rate_unit = ChargingRateUnitEnum::A,
+    };
+    double deltaChanged = 0.2;
+    auto source = ChargingLimitSourceEnum::SO;
+
+    auto resp = handler.handle_external_limits_changed(new_limit, deltaChanged, source, evse_id);
+
+    ASSERT_THAT(resp.has_value(), testing::IsTrue());
+    ASSERT_THAT(resp->chargingLimit.chargingLimitSource, testing::Eq(source));
+    ASSERT_THAT(resp->evseId.has_value(), testing::IsTrue());
+    ASSERT_THAT(resp->evseId.value(), testing::Eq(evse_id));
+}
+
+TEST_F(SmartChargingHandlerTestFixtureV201,
+       K13FR01_HandleClearedChargingLimitRequest_NoTransactionForEvseId_NoRequestCreated) {
```

Comment:
how is this different from `K12FR02_HandleExternalLimitsChanged_LimitChangeSignificanceNotMet_ReturnN...`?

Reply:
The two test are very similar but the `resp` that is being asserted against is from two separate methods `handle_external_limit_cleared` created for K13 and `handle_external_limits_changed` created for K12. They both have independent logic and need to be evaluated even if the outcome is the same.